Documentation: add a CLI reference section

aalemayhu opened the pull request:
Generated via Cobra. Added it to the install file after the agent command reference.

I saw the below warnings from `vndr`, do we need to add them as well?

```
  2017/11/17 09:47:27 WARNING: dependency is not vendored: github.com/cpuguy83/go-md2man/md2man
2017/11/17 09:47:27 WARNING: dependency is not vendored: github.com/russross/blackfriday
```

aalemayhu committed:
cilium: add hidden subcommand for generating command reference

Will be used in the documentation to generate the section on the cilium
CLI reference. Since it is added for this purpose the command is hidden
for now. This can be revisited later.

Also checking in the doc package from cobra and it's dependencies.

Closes: #1586 (Generate command reference via cobra/doc)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
Documentation: generate CLI reference cmdref

The target is not built by default because we don't want to require
having Cilium installed for serving these static files.

The CI should detect changes to the CLI and bark when there is a
mismatch between what is already in the tree compared to new output.

We are not showing this section on the PDF because the links will not
work there because of them being specific to the HTML rendering.

Related: #1586 (Generate command reference via cobra/doc)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
tests: detect changes to the CLI reference

This should hopefully prevent developers from committing user facing CLI
changes without updating the documentation.

Related: #1586 (Generate command reference via cobra/doc)
Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
cilium: sort monitor event types

The Go map does not have any ordering guarantees so when we are for ex.
Generating a command line reference the output can vary on reruns.
Detected while diffing between multiple runs.

While at it remove the eventconfig, since it's not used here anymore.

Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

joestringer reviewed (approved):
Looks neat. From a review perspective and future reference, it's a little easier to review if the autogenerated files are added in an entirely separate commit from the other changes, so we can put more focus into ensuring the handwritten code is correct and just scan across the autogenerated code to check that it looks reasonable. Similarly for vendor/, since a tool generates that as well.

LGTM to me from CLI perspective.

This PR adds a test to `tests/`, please consider filing a ticket to move this over to Ginkgo when we can.

tgraf commented on cilium/cmd/cmdref.go:
```diff
@@ -0,0 +1,43 @@
+var cmdRef = &cobra.Command{
+	Use:   "cmdref",
+	Short: "Generate Cilium command reference",
+	Run: func(cmd *cobra.Command, args []string) {
```
Set `Hidden: true`?
https://godoc.org/github.com/spf13/cobra#Command

aalemayhu replied:
Cool was not aware of that. Makes sense to hide it.

tgraf commented on tests/00-check-cmdref.sh:
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+dir=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
```
Can we run this as part of the build instead or is it too expensive?

aalemayhu replied:
We can do that, but I still, want the runtime check which can prevent someone from forgetting to check-in files.

tgraf replied:
OK, let's do both then

aalemayhu committed:
Makefile: generate the command reference on build

Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
cilium: add hidden subcommand for generating man pages

Adding it now for people who prefer to use man pages. It might also make
it easier for package distributors that want a seperate docs package.

Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>

aalemayhu committed:
cilium: add command prefix to endpoint log example

Makes it easier for copy paste.

Signed-off-by: Alexander Alemayhu <alexander@alemayhu.com>
